// forage/lib/models/forage.players.js
var players,
    Q = require('q'),
    crypto = require('crypto'),
    gravatar = require('gravatar');
    
(function () {
    "use strict";
    
    var Forage = null
    
    players = function () {
        
        var self = this;
        
        this.Player = require('../classes/forage.player');
        
        this.initialize = function () {
            Forage = module.parent.exports;
        };
        
        this.create = function (playerData) {
            var deferred = Q.defer();
            
            playerData = JSON.parse(JSON.stringify(playerData));
            
            playerData = new self.Player(playerData);

            deferred.resolve(playerData);
            
            return deferred.promise;
        };
        
        this.captureByEmail = function (email) {
            var deferred = Q.defer();

            Forage.MongoServer.models.player.findOne({"email": email}, function (err, playerData) {
                deferred.resolve(playerData);
            });
            
            return deferred.promise;
        };
        
        this.validateCredentials = function (email, password) {
            var deferred = Q.defer(),
                playerPassword = crypto.createHash('sha1'),
                player;
            
            self.captureByEmail(email)
                .then(function (playerData) {
                    if (typeof playerData !== 'object'){
                        console.log('invalid email');
                        deferred.reject();
                    } else {
                        password = playerPassword.update(password);
                        if (playerData.password !== password.digest('hex')) {
                            deferred.reject();
                        } else {
                            deferred.resolve(playerData);
                        }
                    }
                });
            
            return deferred.promise;
        };
        
        this.createSession = function (request, player) {
            var deferred = Q.defer();

            if (player.password) {
                delete player.password;
            }
            
            player = JSON.parse(JSON.stringify(player));
            
            if (player.is_admin === -1) {
                delete player.is_admin;
            }
            
            request.session.player = player;
            request.session.player.gravatar = gravatar.url(request.session.player.email, {s: 55, d: 'mm'});
            deferred.resolve(player);            
            return deferred.promise;
        };
        
        this.createNew = function (playerData) {
            var deferred = Q.defer();

            playerData.password = crypto.createHash('sha1').update(playerData.password).digest('hex');
            self.create(playerData)
                .then(function (player) {
                    player.save()
                        .then(function () {
                            deferred.resolve(player);
                        }, console.log);
                }, console.log);
            
            return deferred.promise;
            
        }
        
        this.initialize();
        
    };
    
    players = new players();
    
    // Create the new MongoServer Schema and Model based upong the player object structure
    module.parent.exports.MongoServer.addSchema('player', {
        given_name      : String,
        family_name     : String,
        email           : String,
        password        : String,
        status          : Number,
        is_admin        : Number,
        score           : Number,
        resources       :[Object]
    });
    
}());

module.exports = players;
