// foodlantern/lib/models/foodlantern.recipes.js
var recipes,
    Q = require('q');
    
(function () {
    "use strict";
    
    var FoodLantern = null;
    
    recipes = function () {
        
        var self = this,
            parentGame;
        
        this.Recipe = require('../classes/foodlantern.recipe');
                
        this.listAll = function (request, res) {
            var deferred = Q.defer();
            
            deferred.resolve(self.parentGame.recipes);
            
            return deferred.promise;
        };
        
        this.setParent = function (parent) {
            parentGame = parent;
        };
        
        this.add = function (recipeData) {
            var deferred = Q.defer();

            parentGame.recipes.push(new self.Recipe(recipeData));
            console.log('added recipe to game');
            deferred.resolve();
            return deferred.promise;
        };
        
        this.findById = function (request_ids) {
            var deferred = Q.defer(),
                result = [],
                i;
            
            if (typeof request_ids === 'string') {
                request_ids = request_ids.split(',');
            }
            
            for (i = 0; i < parentGame.recipes.length; i += 1) {
                if (request_ids.indexOf(parentGame.recipes[i]._id) >= 0) {
                    result.push(parentGame.recipes[i]);
                }
            }
            
            deferred.resolve(result);
            return deferred.promise;
        };
        
    };
    
    recipes = new recipes();

    module.parent.parent.parent.exports.MongoServer.addSchema('recipe', {
        name:           String,
        game_id:        String,
        snippet:        String,
        description:    String,
        resources : [
            {
                name:           String,
                description:    String,
                image_url:      String
            }
        ],
        image_uri:      String
    });
    
}());

module.exports = recipes;
