// forage/lib/classes/forage.lantern.js
/*jslint nomen: true */
/*globals require, module, console*/
var Lantern,
    Q = require('q');

(function () {
    "use strict";
    
    var Forage;
    
    Lantern = function (lanternData) {
    
        lanternData = lanternData || {};
        
        var self = this;

        this._id            = '';
        this.game_id        = '';
        this.uid            = '';
        this.name           = '';
        this.description    = '';
        this.notation       = '';
        this.status         = 1;
        this.occupant       = '';
        this.resources      = [];
        this.geolocation    = {
            longitude   : 0,
            latitude    : 0
        };
        
        this.initialize = function (lanternData) {
        
            lanternData = lanternData || {};
            
            Forage = module.parent.parent.exports;
            lanternData = JSON.parse(JSON.stringify(lanternData));
            console.log(lanternData);
            
            for (var d in lanternData) {
                if (self.hasOwnProperty(d) && lanternData.hasOwnProperty(d)) {
                    self[d] = lanternData[d];
                }
            }

            return self;

        };
        
        this.save = function () {
            var deferred = Q.defer(),
            
                newData = {
                    game_id         : self.game_id,
                    uid             : self.uid,
                    name            : self.name,
                    description     : self.description,
                    notation        : self.notation,
                    status          : self.status,
                    occupant        : self.occupant,
                    resources       : self.resources,
                    geolocation     : {
                        longitude       : self.geolocation.longitude,
                        latitude        : self.geolocation.latitude
                    }
                };
            
            if (self._id === '') {
                
                Forage.MongoServer.models.lantern.create(newData, function (err, newLantern) {
                    if (err) {
                        console.log(err)
                    } else {
                        self._id = newLantern._id;
                    }
                    
                    deferred.resolve(self);
                    
                }, console.log);
                
            } else {
                
                Forage.MongoServer.models.lantern.update({_id : self._id}, { $set : newData}, function (err, numberAffected, raw) {
                    if (err) {
                        console.log(err);
                    }

                    // return the player data
                    deferred.resolve(self);

                }, console.log);
                
            }
            
            return deferred.promise;
        };
        
        this.initialize(lanternData);
    };
    
}());

module.exports = Lantern;
